Fixed Cloze Card Preview

prateek-singh-3212 opened the pull request:
## Pull Request template

## Purpose / Description
 This PR Fixes the error which occur while opening the card which is having cloze text from card browser. 

## Fixes
 Fixes #9005

## Approach
I Changed the ordinal to 0 for all the cloze cards.

## How Has This Been Tested?

Tested on Samsung Galaxy M51, android 11 (Samsung One UI 3.1)

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ ] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

prateek-singh-3212 commented:
This PR fixes the error which we are encountering. But I have seen new type of card loading issue while testing. 
I am attaching the video of that issue.


https://user-images.githubusercontent.com/76490368/120347935-56566280-c31a-11eb-82f4-c83068adc83a.mp4

In this I am not able to see the cloze for any card except 1st card i.e It is always showing cloze in first card only.
I have tried to solve this but I am unable to find out how anki is detecting that display 'cloze icon while rewieing 2nd card' ? anyone Please tell me . I hope that i have crearly explained the issue. If not then please tell me.

ShridharGoel commented:
@prateek-singh-3212 Currently AnkiDroid only shows preview for the first card. Refer: #7941.
This is the reason because of which only one cloze symbol is visible.

ShridharGoel commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java:
```diff
@@ -280,6 +280,9 @@ protected Card getCard(Collection col, long cardListIndex) {
             n.setField(i, fieldNames.get(i));
         }
         try {
+            if (model.isCloze()) {
```
A comment can be added here explaining the issue because of which this has been done.

ShridharGoel replied:
Also, this can be moved outside the `try` block.

prateek-singh-3212 replied:
Yes, Sure

ShridharGoel commented:
@prateek-singh-3212 Are you planning to add an unit test for this?

david-allison-1 commented:
This is from the previewer (card browser - now editor), rather than from the note editor directly. Two similar, but unrelated issues

prateek-singh-3212 commented:
> @prateek-singh-3212 Are you planning to add an unit test for this?

Yes, I am thinking that unit test will be great for this issue.

david-allison-1 commented:
Tested manually and it's good. I'd like to see this in 2.15.3 (with the caveat that the change to get this fully working will be in 2.15.4).

Happy to see a unit test in a separate PR to allow this to get in, or a unit test here if you have time today.

david-allison-1 reviewed (approved):
See comment: unit test to follow up would be great, but would like to see this in 2.15.3 if possible. Happy to see a test here, or as a follow up PR if this gets merged before a test is added

(Added the comment so an issue can be created directly from it)

prateek-singh-3212 commented:
I have implemented changes as suggested by @ShridharGoel .

-  Implemented if statement outside the try block.
- Added Comment which describes the reason to implement this.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java:
```diff
@@ -279,6 +279,15 @@ protected Card getCard(Collection col, long cardListIndex) {
+
+        /**
+         * This is used when the card type is cloze. We have to set the ordinal to 0 as we are
+         * setting the 0 for all cards of a particular note type in are database.
```
Also sorry for the review

Comment the why, not the what. Maybe move the first sentence to the header of the method

```
This method generates a note from a sample model, or fails if invalid. It does not currently have knowledge of field content

A cloze uses the same model. Its content (not provided in params) determines validity of an ordinal
```

prateek-singh-3212 replied:
Done

david-allison-1 reviewed (approved):
A thing of beauty!

mikehardy reviewed (approved):
Just in time for 2.15.3 :-)
